File: September Challenge/Week 3/Java/Day_17/RobotBoundedInCircle.java
// Problem : https://leetcode.com/problems/robot-bounded-in-circle/
// Solution:

class Solution {
    public boolean isRobotBounded(String instructions) {
        int x=0;
        int y=0;
        char dirn='N';
        
        for(char c:instructions.toCharArray()){
            if(c=='G'){
                if(dirn=='N') y++;
                else if(dirn=='S') y--;
                else if(dirn=='E') x++;
                else x--;
            }else if(c=='L'){
                if(dirn=='N') dirn='W';
                else if(dirn=='S') dirn='E';
                else if(dirn=='E') dirn='N';
                else dirn='S';
            }else{
                if(dirn=='N') dirn='E';
                else if(dirn=='S') dirn='W';
                else if(dirn=='E') dirn='S';
                else dirn='N';
            }
        }
        
        if((dirn!='N')||(x==0 && y==0)){
            return true;
        }
        return false;
    }
}

File: September Challenge/Week 3/C++/Day_17/RobotBoundedInCircle.cpp
class Solution {
public:

	bool isRobotBounded(string instructions) {
    
    // hash map giving the new directions in which we will be facing
    unordered_map<char,pair<char,char>> dir{{'N',{'W','E'}},{'S',{'E','W'}},{'E',{'N','S'}},{'W',{'S','N'}}};
    
    // hashmap giving the change in the cordinates after moving in particular direction
    unordered_map<char,pair<int,int>> moves{{'N',{0,1}},{'S',{0,-1}},{'E',{1,0}},{'W',{-1,0}}};
    
    // initial position of robot
    int i=0,j=0;
    
    // initial facing direction
    char facing = 'N';
    
    
    for(auto c:instructions)
    {
        if(c=='L')
        {
            facing=dir[facing].first;
        }
        
        else if(c=='R')
        {
            facing=dir[facing].second;
        }
        
        else if(c=='G')
        {
            i+=moves[facing].first;
            j+=moves[facing].second;
        }
            
    }
    
    if(i==0 and j==0)
    {
        return true;
    }
    
    return facing!='N';
    
    
}
};
